### PR TITLE
Musig2: Update to 1.0.0.rc2

### PR DESCRIPTION
**Pull request by sputn1ck**
This PR adds the changes from https://github.com/jonasnick/bips/pull/74 which adresses the risk described in https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-October/021000.html .

Will leave it as draft, until the bip PR is merged.

The `bench_test.go/BenchmarkPartialVerify` are failing, however they also seem to be failing on master.

### CODE REVIEW

**Comment by sputn1ck** [comment]
https://github.com/jonasnick/bips/pull/74 is now merged, now new changes to the reference code have been made sind my last commits

File: btcec/schnorr/musig2/context.go [hunk 1]
```diff
@@ -240,6 +240,7 @@ func NewContext(signingKey *btcec.PrivateKey, shouldSort bool,
 		if opts.earlyNonce {
 			var err error
 			ctx.sessionNonce, err = GenNonces(
+				WithPublicKey(ctx.pubKey),
```
Comment (guggero) [inline btcec/schnorr/musig2/context.go:243]: Ah nice, so on the API level this change shouldn't have any impact at all.

File: btcec/schnorr/musig2/data/sig_agg_vectors.json [hunk 2]
```diff
@@ -6,81 +6,146 @@
...
         {
-            "aggnonce": "02BC34CDF6FA1298D7B6A126812FAD0739005BC44E45C21276EEFE41AAF841C86F03F3562AED52243BB99F43D1677DB59F0FEFB961633997F7AC924B78FBD0B0334F",
-            "nonce_indices": [0, 1],
-            "key_indices": [0, 1],
+            "aggnonce": "0341432722C5CD0268D829C702CF0D1CBCE57033EED201FD335191385227C3210C03D377F2D258B64AADC0E16F26462323D701D286046A2EA93365656AFD9875982B",
+            "nonce_indices": [
+                0,
+                1
+            ],
+            "key_indices": [
```
Comment (guggero) [inline btcec/schnorr/musig2/data/sig_agg_vectors.json:40]: nit: avoid these formatting changes (probably done by the IDE) to keep the diff smaller?

File: btcec/schnorr/musig2/nonces.go [hunk 3]
```diff
@@ -34,6 +35,10 @@ var (
 	NonceGenTag = []byte("MuSig/nonce")
 
 	byteOrder = binary.BigEndian
+
+	// ErrPubkeyInvalid is returned when the pubkey of the WithPublicKey
+	// option is not passed or of invalid length.
+	ErrPubkeyInvalid = errors.New("nonce generation requires a valid pubkey")
```
Comment (guggero) [inline btcec/schnorr/musig2/nonces.go:41]: nit: use `PubKey` capitalization in name?

File: btcec/schnorr/musig2/nonces.go [hunk 4]
```diff
@@ -142,6 +151,14 @@ func WithCustomRand(r io.Reader) NonceGenOption {
 	}
 }
 
+// WithPublicKey is the mandatory public key that will be mixed into the nonce
```
Comment (guggero) [inline btcec/schnorr/musig2/nonces.go:154]: Not sure if it's obvious from the context or whether we should add "the mandatory public key that corresponds to the signer's private key" or something to that effect to the comment?

File: btcec/schnorr/musig2/sign.go [hunk 5]
```diff
@@ -38,6 +38,10 @@ var (
 	// ErrSecretNonceZero is returned when a secret nonce is passed in a
 	// zero.
 	ErrSecretNonceZero = fmt.Errorf("secret nonce is blank")
+
+	// ErrSecNoncePubkey is returned when the signing key does not match the
```
Comment (guggero) [inline btcec/schnorr/musig2/sign.go:42]: nit: capitalization of PubKey and missing full stop.

**Review by guggero — APPROVED** [review]
Thanks for the update, LGTM :tada:

**Review by Roasbeef — APPROVED** [review]
LGTM 🥟

Will make another PR to fix the issue with the benchmark.

**Pull request merged**